File: Sources/grostat/Alerts.swift
import Foundation
import GrostatShared

struct AlertChecker {
    let ratedPowerW: Int
    let criticalV: Double
    let criticalClearV: Double
    private let stateFile: String

    init(config: Config) {
        self.ratedPowerW = config.ratedPowerW
        self.criticalV = config.alertCriticalV
        self.criticalClearV = config.alertCriticalV - 2  // 253 → 251 hysteresis
        let dir = (config.resolvedDbPath as NSString).deletingLastPathComponent
        self.stateFile = "\(dir)/alert_state.json"
    }

    func evaluate(_ reading: InverterReading) -> String {
        let previous = readState()
        var current = previous

        // Production state
        let prodState = productionState(reading)
        if prodState != previous.productionState {
            current.productionState = prodState
            let (title, message) = productionNotification(prodState, reading)
            notify(title: title, message: message)
        }

        // Voltage with hysteresis
        let v = reading.vmaxPhase
        if !previous.voltageCritical && v >= criticalV {
            current.voltageCritical = true
            notify(
                title: "grostat CRITICAL",
                message: "\(f1(v))V — grid voltage exceeded \(f1(criticalV))V!",
                subtitle: "R:\(f1(reading.vacrPhase))V S:\(f1(reading.vacsPhase))V T:\(f1(reading.vactPhase))V"
            )
        } else if previous.voltageCritical && v < criticalClearV {
            current.voltageCritical = false
            notify(
                title: "grostat OK",
                message: "Voltage back to normal: \(f1(v))V",
                subtitle: "R:\(f1(reading.vacrPhase))V S:\(f1(reading.vacsPhase))V T:\(f1(reading.vactPhase))V"
            )
        }

        if current != previous {
            writeState(current)
        }

        // Return alert level for DB column
        if current.voltageCritical { return "CRITICAL" }
        if prodState == "fault" { return "FAULT" }
        return ""
    }

    // MARK: - Production state (mirrors GrostatBar InverterState)

    private func productionState(_ r: InverterReading) -> String {
        if r.isOffline { return "offline" }
        if r.status == 3 { return "fault" }
        if r.status == 0 || r.ppv == 0 { return "sleep" }
        let onFireThreshold = Double(ratedPowerW) * 0.7
        if r.ppv >= onFireThreshold { return "onFire" }
        let cloudyThreshold = Double(ratedPowerW) * 0.4
        if r.ppv < cloudyThreshold { return "cloudy" }
        return "producing"
    }

    private func productionNotification(_ state: String, _ r: InverterReading) -> (title: String, message: String) {
        let kw = String(format: "%.1f", r.ppv / 1000.0)
        switch state {
        case "sleep":
            return ("grostat", "Inverter sleeping")
        case "cloudy":
            return ("grostat", "Low production: \(kw) kW")
        case "producing":
            return ("grostat", "Producing: \(kw) kW")
        case "onFire":
            return ("grostat", "Full power: \(kw) kW")
        case "fault":
            return ("grostat FAULT", "Inverter fault (type \(r.faultType))")
        case "offline":
            return ("grostat", "Inverter offline")
        default:
            return ("grostat", "State: \(state)")
        }
    }

    // MARK: - State persistence

    private struct AlertState: Codable, Equatable {
        var productionState: String = ""
        var voltageCritical: Bool = false
    }

    private func readState() -> AlertState {
        guard let data = FileManager.default.contents(atPath: stateFile),
              let state = try? JSONDecoder().decode(AlertState.self, from: data)
        else { return AlertState() }
        return state
    }

    private func writeState(_ state: AlertState) {
        guard let data = try? JSONEncoder().encode(state) else { return }
        try? data.write(to: URL(fileURLWithPath: stateFile))
    }

    // MARK: - Notifications

    private func notify(title: String, message: String, subtitle: String = "") {
        #if os(macOS)
            var script = "display notification \"\(message)\" with title \"\(title)\""
            if !subtitle.isEmpty {
                script += " subtitle \"\(subtitle)\""
            }
            let process = Process()
            process.executableURL = URL(fileURLWithPath: "/usr/bin/osascript")
            process.arguments = ["-e", script]
            process.standardOutput = FileHandle.nullDevice
            process.standardError = FileHandle.nullDevice
            try? process.run()
            process.waitUntilExit()
        #endif
    }

    private func f1(_ v: Double) -> String {
        String(format: "%.1f", v)
    }
}

File: src/grostat/alerts.py
"""Two-tier voltage alert system with macOS notifications."""

import logging
import platform
import subprocess

from grostat.config import Settings
from grostat.models import InverterReading

logger = logging.getLogger("grostat")


class AlertChecker:
    def __init__(self, settings: Settings) -> None:
        self.warning_v = settings.alert_warning_v
        self.critical_v = settings.alert_critical_v
        self._is_macos = platform.system() == "Darwin"

    def evaluate(self, reading: InverterReading) -> str:
        """Evaluate reading and return alert level: '', 'WARNING', or 'CRITICAL'."""
        v = reading.vmax_phase

        if v >= self.critical_v:
            level = "CRITICAL"
            logger.critical(
                "CRITICAL: vmax_phase=%.1fV (>= %.1fV) on phases R:%.1f S:%.1f T:%.1f",
                v,
                self.critical_v,
                reading.vacr_phase,
                reading.vacs_phase,
                reading.vact_phase,
            )
            self._notify(
                f"CRITICAL: {v:.1f}V (grid voltage exceeded!)",
                f"R:{reading.vacr_phase}V S:{reading.vacs_phase}V T:{reading.vact_phase}V",
                title="grostat CRITICAL",
            )
        elif v >= self.warning_v:
            level = "WARNING"
            logger.warning(
                "WARNING: vmax_phase=%.1fV (>= %.1fV) on phases R:%.1f S:%.1f T:%.1f",
                v,
                self.warning_v,
                reading.vacr_phase,
                reading.vacs_phase,
                reading.vact_phase,
            )
            self._notify(
                f"WARNING: {v:.1f}V (approaching grid limit)",
                f"R:{reading.vacr_phase}V S:{reading.vacs_phase}V T:{reading.vact_phase}V",
                title="grostat",
            )
        else:
            level = ""

        return level

    def _notify(self, message: str, subtitle: str, title: str = "grostat") -> None:
        if not self._is_macos:
            return
        try:
            script = f'display notification "{message}" with title "{title}" subtitle "{subtitle}"'
            subprocess.run(
                ["osascript", "-e", script],
                capture_output=True,
                timeout=5,
            )
        except Exception:
            pass  # notification is best-effort
